docs/source/user/vectorize.rst

```diff
@@ -101,6 +101,9 @@ reduction, accumulation or broadcasting.  Using the example above::
 .. seealso::
    `Standard features of ufuncs <http://docs.scipy.org/doc/numpy/reference/ufuncs.html#ufunc>`_ (NumPy documentation).
 
+.. note::
+   Some features of ufuncs (like reduction and accumulation) are not supported
+   in compiled code.
```

Comment:
I'm not sure that this sufficiently clarifies the situation? As I understand it the issue is that already existing NumPy ufuncs don't necessarily have their features available in `nopython` mode compiled code, but `ufuncs` (actually `DUFunc`s) created from `@vectorize` do? I think `@vectorize` with a signature will also need checking as that too is a slightly different code path. And further the use of all these inside/outside of `nopython` mode. Perhaps after this investigation a statement with intent similar to the above can be constructed more easily/accurately?

Reply:
I don't think DUFuncs have those features available.

### `@vectorize`

Signature-less `@vectorize` doesn't seem to have ufunc attributes (I thought this worked at one point, but could be misremembering):

```python
from numba import vectorize, njit
import numpy as np

@vectorize
def foo(a, b):
    return a + b

foo.reduce(np.arange(10))
```

Traceback (similar if you try to call from a `@njit` decorated function):

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-1a307d656cc6> in <module>
      6     return a + b
      7 
----> 8 foo.reduce(np.arange(10))

TypeError: No loop matching the specified signature and casting was found for ufunc foo
```


### `@vectorize(sig)`

With a signature, it does get ufunc methods. Those don't work in compiled functions:

```python
@vectorize(['int64(int64, int64)'])
def foo(a, b):
    return a + b

foo.reduce(np.arange(10))  # 45

@numba.njit
def bar(a):
    return foo.reduce(a)
    
bar(np.arange(10))
```



<details>
<summary>Traceback: </summary>


```python
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-5-2fcd7b37ff2e> in <module>
----> 1 bar(np.arange(10))

/usr/local/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    374                 e.patch_message(msg)
    375 
--> 376             error_rewrite(e, 'typing')
    377         except errors.UnsupportedError as e:
    378             # Something unsupported is present in the user code, add help info

/usr/local/lib/python3.7/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    341                 raise e
    342             else:
--> 343                 reraise(type(e), e, None)
    344 
    345         argtypes = []

/usr/local/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'reduce' of type Function(<numba._DUFunc 'foo'>)

File "<ipython-input-4-e5cf30853351>", line 3:
def bar(a):
    return foo.reduce(a)
    ^

[1] During: typing of get attribute at <ipython-input-4-e5cf30853351> (3)

File "<ipython-input-4-e5cf30853351>", line 3:
def bar(a):
    return foo.reduce(a)
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/latest/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

</details>

### Numpy ufunc

```python
@njit
def bar(a):
    return np.add.reduce(a)

bar(np.arange(10))
```

<details>
<summary> Traceback: </summary>

```python
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-6-4468ceee1e62> in <module>
      3     return np.add.reduce(a)
      4 
----> 5 bar(np.arange(10))

/usr/local/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    374                 e.patch_message(msg)
    375 
--> 376             error_rewrite(e, 'typing')
    377         except errors.UnsupportedError as e:
    378             # Something unsupported is present in the user code, add help info

/usr/local/lib/python3.7/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    341                 raise e
    342             else:
--> 343                 reraise(type(e), e, None)
    344 
    345         argtypes = []

/usr/local/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'reduce' of type Function(<ufunc 'add'>)

File "<ipython-input-6-4468ceee1e62>", line 3:
def bar(a):
    return np.add.reduce(a)
    ^

[1] During: typing of get attribute at <ipython-input-6-4468ceee1e62> (3)

File "<ipython-input-6-4468ceee1e62>", line 3:
def bar(a):
    return np.add.reduce(a)
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/latest/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

</details>

Reply:
Following up on my previous curt reply:

I'm not sure I've seen ufunc methods (e.g. `{ufunc}.reduce`) work inside of a compiled function. I've definitely tried it, but after your referenced comment figured the reason I hadn't seen it work was that it was currently unsupported.

I've provided the cases I've check above – I've also checked `.accumulate`, so I don't think the issue is `reduce` specific. If there are ufunc method calls that do work, that'd be useful to see. The main reason I wanted to update the documentation was it took me a while to try and figure out if this was a bug, or just unsupported behavior.

Reply:
Thanks for looking into this and checking all these examples. What you present seems accurate, I can reproduce. Do you think it is therefore accurate to say:

> ufuncs methods (like reduction and accumulation) are not supported if called within a JIT compiled function.

as an update to the statement made in this PR?

Reply:
It would be good to be more precise about unsupported features. I'm just not sure what range of features is supported in compiled code. For example, I just noticed ufunc attribute access (i.e. `ufunc.nin`) doesn't work in compiled code.

Reply:
I don't think any of the stuff "dangling" off a ufunc is likely to work in compiled code. I don't recall all the details immediately, but it's to do with how ufuncs are generated and their relationship to NumPy's ufunc mechanism/wrappers.

Reply:
I guess I'm not sure I know what to consider as "core features" of ufuncs. Would you say it's just the broadcasting? I just tried a few keyword arguments (`out`, `axis`) and they didn't seem to work. How about this:

> Only the broadcasting features of ufuncs are supported in compiled code.

Reply:
hmmm, `out` should work, but that aside, I think the statement above is valid.

Reply:
Could you provide a case of `out` working? I get errors from this (even if I type `foo`):

```python
from numba import vectorize, njit
import numpy as np

@vectorize
def foo(x, y):
    return x + y

@njit
def bar(x, y, z):
    foo(x, y, out=z)

a = np.ones(10)
b = np.ones(10)
out = np.zeros(10)

bar(a, b, out)
```

Reply:
`out` works but you can't pass by key-word:
```python
from numba import vectorize, njit
import numpy as np

@vectorize
def foo(x, y):
    return x + y

@njit
def bar(x, y, z):
    foo(x, y, z)

a = np.ones(10)
b = np.ones(10)
out = np.zeros(10)

bar(a, b, out)
print(out)
```
IIRC think there's an open issue about this but I can't find it :)